File: zCLI/subsystems/zLoader/zLoader_modules/pinned_cache.py
# zCLI/subsystems/zLoader/zLoader_modules/pinned_cache.py

"""User-loaded aliases cache - never auto-evicts, highest priority."""

import time

class PinnedCache:
    """Pinned cache for user-loaded aliases (parsed schemas only, no auto-eviction)."""

    def __init__(self, session, logger):
        """Initialize pinned cache."""
        self.session = session
        self.logger = logger
        self._ensure_namespace()

    def _ensure_namespace(self):
        """Ensure pinned_cache namespace exists in session."""
        if "zCache" not in self.session:
            self.session["zCache"] = {}

        if "pinned_cache" not in self.session["zCache"]:
            self.session["zCache"]["pinned_cache"] = {}

    def load_alias(self, alias_name, parsed_schema, zpath):
        """Load alias (from load --as command)."""
        try:
            cache = self.session["zCache"]["pinned_cache"]
            key = f"alias:{alias_name}"

            cache[key] = {
                "data": parsed_schema,
                "zpath": zpath,
                "type": "schema",
                "loaded_at": time.time()
            }

            self.logger.info("[PinnedCache] Alias loaded: $%s → %s", alias_name, zpath)

        except Exception as e:
            self.logger.error("[PinnedCache ERROR] %s - %s", alias_name, e)

        return parsed_schema

    def get_alias(self, alias_name):
        """Get alias by name."""
        try:
            cache = self.session["zCache"]["pinned_cache"]
            key = f"alias:{alias_name}"

            if key in cache:
                entry = cache[key]
                self.logger.debug("[PinnedCache HIT] $%s", alias_name)
                return entry.get("data")

            return None

        except Exception as e:
            self.logger.debug("[PinnedCache ERROR] %s - %s", alias_name, e)
            return None

    def has_alias(self, alias_name):
        """Check if alias exists."""
        try:
            key = f"alias:{alias_name}"
            return key in self.session["zCache"]["pinned_cache"]
        except Exception:
            return False

    def remove_alias(self, alias_name):
        """Remove specific alias."""
        try:
            cache = self.session["zCache"]["pinned_cache"]
            key = f"alias:{alias_name}"

            if key in cache:
                del cache[key]
                self.logger.info("[PinnedCache] Removed: $%s", alias_name)
                return True
            return False
        except Exception as e:
            self.logger.error("[PinnedCache ERROR] %s - %s", alias_name, e)
            return False

    def clear(self, pattern=None):
        """Clear pinned resources (optionally by pattern)."""
        try:
            cache = self.session["zCache"]["pinned_cache"]

            if pattern:
                # Clear matching keys
                pattern_str = pattern.replace("*", "")
                keys_to_delete = [k for k in cache.keys() if pattern_str in k]
                for key in keys_to_delete:
                    del cache[key]
                self.logger.info(
                    "[PinnedCache] Removed %d aliases matching '%s'",
                    len(keys_to_delete), pattern
                )
                return len(keys_to_delete)
            else:
                # Clear all
                count = len(cache)
                cache.clear()
                self.logger.info("[PinnedCache] Removed all %d aliases", count)
                return count

        except Exception as e:
            self.logger.error("[PinnedCache ERROR] clear - %s", e)
            return 0

    def list_aliases(self):
        """List all aliases."""
        try:
            cache = self.session["zCache"]["pinned_cache"]

            aliases = []
            for key, entry in cache.items():
                if key.startswith("alias:"):
                    alias_name = key.replace("alias:", "")
                    aliases.append({
                        "name": alias_name,
                        "zpath": entry.get("zpath"),
                        "type": entry.get("type"),
                        "loaded_at": entry.get("loaded_at"),
                        "age": time.time() - entry.get("loaded_at", time.time())
                    })

            return aliases

        except Exception as e:
            self.logger.error("[PinnedCache ERROR] list - %s", e)
            return []

    def get_info(self, alias_name):
        """Get metadata about an alias."""
        try:
            cache = self.session["zCache"]["pinned_cache"]
            key = f"alias:{alias_name}"

            if key in cache:
                entry = cache[key]
                return {
                    "name": alias_name,
                    "zpath": entry.get("zpath"),
                    "type": entry.get("type"),
                    "loaded_at": entry.get("loaded_at"),
                    "age": time.time() - entry.get("loaded_at", time.time()),
                    "size": len(str(entry.get("data", "")))
                }

            return None

        except Exception as e:
            self.logger.error("[PinnedCache ERROR] get_info - %s", e)
            return None

File: zCLI/subsystems/zComm/zComm_modules/zBifrost/bifrost_client_modules/pinned_cache.js
/**
 * ═══════════════════════════════════════════════════════════════
 * PinnedCache - User aliases/bookmarks (never auto-evicts)
 * ═══════════════════════════════════════════════════════════════
 * 
 * Mirrors zLoader's pinned_cache for user-loaded aliases.
 * Resources pinned by user are never automatically evicted.
 * Persists to IndexedDB/localStorage for cross-session use.
 * 
 * @version 1.5.4
 * @author Gal Nachshon
 * @license MIT
 */

class PinnedCache {
    constructor(options = {}) {
        this.cache = new Map();
        this.storage = options.storage;
        this.debug = options.debug || false;
        
        // Load from persistent storage on init
        if (this.storage) {
            this._loadFromStorage();
        }
    }
    
    /**
     * Load/pin a resource with alias
     * @param {string} alias - User-defined alias
     * @param {any} data - Data to pin
     * @param {Object} metadata - Optional metadata
     * @returns {Promise<any>} The pinned data
     */
    async load(alias, data, metadata = {}) {
        const entry = {
            data: data,
            loadedAt: Date.now(),
            metadata: metadata
        };
        
        this.cache.set(alias, entry);
        
        if (this.storage) {
            await this.storage.set(`pinned:${alias}`, entry);
        }
        
        if (this.debug) {
            console.log(`[PinnedCache] LOAD ${alias}`, metadata);
        }
        
        return data;
    }
    
    /**
     * Get pinned resource by alias
     * @param {string} alias - User-defined alias
     * @returns {any} Pinned data or null
     */
    get(alias) {
        const entry = this.cache.get(alias);
        if (this.debug && entry) {
            console.log(`[PinnedCache] HIT ${alias}`);
        }
        return entry ? entry.data : null;
    }
    
    /**
     * Check if alias exists
     * @param {string} alias - User-defined alias
     * @returns {boolean} True if alias exists
     */
    has(alias) {
        return this.cache.has(alias);
    }
    
    /**
     * Remove/unpin specific alias
     * @param {string} alias - User-defined alias
     */
    async remove(alias) {
        this.cache.delete(alias);
        
        if (this.storage) {
            await this.storage.delete(`pinned:${alias}`);
        }
        
        if (this.debug) {
            console.log(`[PinnedCache] REMOVE ${alias}`);
        }
    }
    
    /**
     * Clear all pinned resources
     */
    async clear() {
        const size = this.cache.size;
        this.cache.clear();
        
        if (this.storage) {
            const keys = await this.storage.getAllKeys('pinned:');
            await Promise.all(keys.map(k => this.storage.delete(k)));
        }
        
        if (this.debug) {
            console.log(`[PinnedCache] CLEAR (${size} aliases removed)`);
        }
    }
    
    /**
     * List all pinned aliases
     * @returns {Array} Array of alias info
     */
    listAll() {
        return Array.from(this.cache.entries()).map(([alias, entry]) => ({
            alias,
            loadedAt: entry.loadedAt,
            metadata: entry.metadata,
            age: Date.now() - entry.loadedAt
        }));
    }
    
    /**
     * Get cache statistics
     * @returns {Object} Cache stats
     */
    getStats() {
        return {
            size: this.cache.size,
            aliases: Array.from(this.cache.keys())
        };
    }
    
    /**
     * Load pinned resources from persistent storage
     * @private
     */
    async _loadFromStorage() {
        if (!this.storage) return;
        
        try {
            const entries = await this.storage.getAll('pinned:');
            for (const [key, entry] of entries) {
                const alias = key.replace('pinned:', '');
                this.cache.set(alias, entry);
            }
            
            if (this.debug && entries.length > 0) {
                console.log(`[PinnedCache] Loaded ${entries.length} aliases from storage`);
            }
        } catch (e) {
            console.warn('[PinnedCache] Failed to load from storage:', e);
        }
    }
}

// Export for module systems
if (typeof module !== 'undefined' && module.exports) {
    module.exports = PinnedCache;
}

// Export for browser global
if (typeof window !== 'undefined') {
    window.PinnedCache = PinnedCache;
}
